### PR TITLE
Use `argh` to handle command line options

### PR DESCRIPTION
**Pull request by liuw**
This reduces the binary size by a few hundred kilobytes and shrinks the number of dependencies.

~Draft since this is WIP and I want to test if the conversion is okay.~

### CODE REVIEW

**Comment by rbradford** [comment]
@liuw This is excellent. Lets go ahead and do this for the main binary?

**Review by sboeuf — COMMENTED** [review]
I agree this looks great, not too intrusive!

**Comment by liuw** [comment]
I have an early patch to convert the main binary.

The saving is about 0.3M.

A few caveats:

1. `-vvvvv` needs to be written as `-v -v -v -v -v`.
2. `--disk DISK1 DISK2`  and others need to be written as `--disk DISK1 --disk DISK2`.
3. `--option=value` should be written as `--option value`.
4. Need to pull some code from argh for unit tests since it does not have a helper to parse arbitrary strings.
5. May want to provide a separate document for syntax.

**Comment by sboeuf** [comment]
> I have an early patch to convert the main binary.
> 
> The saving is about 0.3M.

It's not as big as expected IIUC, but we're still getting the benefit of reducing the list of crates we depend on.

> 
> A few caveats:
> 
>     1. `-vvvvv` needs to be written as `-v -v -v -v -v`.
> 
>     2. `--disk DISK1 DISK2`  and others need to be written as `--disk DISK1 --disk DISK2`.

I consider this one as the change that will have the bigger impact on our documentation, tests, etc...
That being said, this syntax might feel more natural for our users.

> 
>     3. `--option=value` should be written as `--option value`.

I'm not worried about this one given I don't think we were using it that much.

> 
>     4. Need to pull some code from argh for unit tests since it does not have a helper to parse arbitrary strings.
> 
>     5. May want to provide a separate document for syntax.

**Comment by liuw** [comment]
@sboeuf @rbradford The main binary is done. I also updated all documents in tree.

On a 32-core VM.

Before: dependencies 139, build time 57.25s.
After: dependencies 130, build time 54.66s.

Right now the help strings are heavily simplified. I can copy&paste some text from `vmm/src/configrs` if necessary. I'm just not sure if that's the best thing to do right now, because it can easily go out of sync.

```
Usage: cloud-hypervisor [--cpus <cpus>] [--platform <platform>] [--memory <memory>] [--memory-zone <memory-zone...>] [--firmware <firmware>] [--kernel <kernel>] [--initramfs <initramfs>] [--cmdline <cmdline>] [--disk <disk...>] [--net <net...>] [--rng <rng>] [--balloon <balloon>] [--fs <fs...>] [--pmem <pmem...>] [--serial <serial>] [--console <console>] [--device <device...>] [--user-device <user-device...>] [--vdpa <vdpa...>] [--vsock <vsock>] [--numa <numa...>] [--watchdog] [-v] [--log-file <log-file>] [--api-socket <api-socket>] [--event-monitor <event-monitor>] [--restore <restore>] [--seccomp <seccomp>] [--tpm <tpm>] [--sgx-epc <sgx-epc...>] [-V]

Launch a cloud-hypervisor VMM.

Options:
  --cpus            specify CPU configuration
  --platform        specify platform configuration
  --memory          specify memory configuration
  --memory-zone     specify memory zone configuration
  --firmware        path to firmware that is loaded in an architectural specific
                    way
  --kernel          path to kernel or firmware that supports a PVH entry point
                    or architecture equivalent
  --initramfs       path to initramfs image
  --cmdline         kernel command line
  --disk            specify disk configuration
  --net             specify network interface configuration
  --rng             specify random number generator configuration
  --balloon         specify balloon configuration
  --fs              specify virtiofs configuration
  --pmem            specify pmem configuration
  --serial          specify serial port configuration
  --console         specify console configuration
  --device          specify device configuration
  --user-device     specify user device configuration
  --vdpa            specify VDPA configuration
  --vsock           specify vsock configuration
  --numa            specify NUMA configuration
  --watchdog        enable virtio-watchdog
  -v, --verbosity   set the level of debugging output
  --log-file        path to log file
  --api-socket      path to HTTP API socket (UNIX domain socket)
  --event-monitor   file to report events on
  --restore         configuration to restore a guest
  --seccomp         seccomp configuration (true, false or log)
  --tpm             specify TPM configuration
  --sgx-epc         specify SGx EPC configuration
  -V, --version     print version information
  --help            display usage information
```

File: Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,7 @@ strip = true
 [dependencies]
 anyhow = "1.0.68"
 api_client = { path = "api_client" }
+argh = "0.1.9"
```
Comment (sboeuf) [inline Cargo.toml:29]: You forgot to remove the clap crate here.

**Review by sboeuf — COMMENTED** [review]
Looks good

File: Cargo.toml (hunk 1 above)
Comment (sboeuf) [reply Cargo.toml:29]: Ah yes of course. Both `ch-remote` and `cloud-hypervisor` binaries depend on this `Cargo.toml`.

**Comment by rbradford** [comment]
Cool, so it looks like we need to document what the parameters look like somewhere then if they're not in the `--help` output for `cloud-hypervisor` or `ch-remote`. Any thoughts?

**Comment by liuw** [comment]
> Cool, so it looks like we need to document what the parameters look like somewhere then if they're not in the `--help` output for `cloud-hypervisor` or `ch-remote`. Any thoughts?

There are documents in tree, but for a deployment scenario those are not going to be available.

The simplest thing to do is to _move_ all the SYNTAX constants in vmm/src/config.rs to src/main.rs. That way there is still only one place to update for a given option's syntax.

**Comment by rbradford** [comment]
> > Cool, so it looks like we need to document what the parameters look like somewhere then if they're not in the `--help` output for `cloud-hypervisor` or `ch-remote`. Any thoughts?
> 
> There are documents in tree, but for a deployment scenario those are not going to be available.
> 
> The simplest thing to do is to _move_ all the SYNTAX constants in vmm/src/config.rs to src/main.rs. That way there is still only one place to update for a given option's syntax.

But `ch-remote` shares e.g. the network config statement - would it still be duplicated?

**Comment by liuw** [comment]
> 
> Actually, I may have a better idea with using `#doc` attribute. With that we may be able to put syntax into separate files and then do `#[doc = !include_str(XXX.syntax)]`.

Nope. A quick test shows that `#[doc = "abc"]` works but `#[doc = include_str!("file")]` does not.

**Comment by liuw** [comment]
We may also add a switch `--print-configuration-syntax` to print them out.

**Comment by rbradford** [comment]
> Unfortunately yes. We can perhaps say "the syntax is the same as Cloud Hypervisor's syntax" to point people to the main binary.

This seems a perfectly good solution to me!

**Review by likebreath — COMMENTED** [review]
Thanks for such a comprehensive work. This PR looks great. 

Only one nit: can you please add the version command for `ch-remote`, `vhost-user-net`, and `vhost-user-blk`? I noticed the `ch-remote` does not have a version command before this change. But I think it is good to add the command, given ch-remote is included in our release.

**Comment by liuw** [comment]
> Thanks for such a comprehensive work. This PR looks great.
> 
> Only one nit: can you please add the version command for `ch-remote`, `vhost-user-net`, and `vhost-user-blk`? I noticed the `ch-remote` does not have a version command before this change. But I think it is good to add the command, given ch-remote is included in our release.

Adding one for `ch-remote` is fine. For the other two, they don't have `build.rs`. I can do another PR for adding that.

File: vhost_user_net/src/main.rs [hunk 2]
```diff
@@ -15,13 +15,27 @@ struct TopLevel {
...
     let toplevel: TopLevel = argh::from_env();
 
-    start_net_backend(&toplevel.backend_command);
+    if toplevel.version {
+        println!("{} {}", env!("CARGO_BIN_NAME"), env!("BUILT_VERSION"));
+        return;
+    }
+
+    if toplevel.backend_command.is_none() {
+        println!("Please specify --block-backend");
```
Comment (rbradford) [inline vhost_user_net/src/main.rs:36]: `--block-backend` ?

**Review by rbradford — APPROVED** [review]

**Pull request merged**